autogpt_platform/backend/backend/server/integrations/creds_manager.py

```diff
@@ -0,0 +1,126 @@
+import logging
+from contextlib import contextmanager
+from datetime import datetime
+
+from autogpt_libs.supabase_integration_credentials_store import (
+    Credentials,
+    SupabaseIntegrationCredentialsStore,
+)
+from autogpt_libs.utils.synchronize import KeyedMutex
+
+from backend.integrations.oauth import HANDLERS_BY_NAME, BaseOAuthHandler
+from backend.util.settings import Settings
+
+from .utils import get_supabase
+
+logger = logging.getLogger(__name__)
+settings = Settings()
+
+
+class IntegrationCredentialsManager:
+    _locks = KeyedMutex()
+
+    def __init__(self):
+        self.store = SupabaseIntegrationCredentialsStore(get_supabase())
+
+    def create(self, user_id: str, credentials: Credentials) -> None:
+        return self.store.add_creds(user_id, credentials)
+
+    def exists(self, user_id: str, credentials_id: str) -> bool:
+        return self.store.get_creds_by_id(user_id, credentials_id) is not None
+
+    def get(
+        self, user_id: str, credentials_id: str, lock: bool = True
+    ) -> Credentials | None:
+        credentials = self.store.get_creds_by_id(user_id, credentials_id)
+        if not credentials:
+            return None
+
+        # Refresh OAuth credentials if needed
+        if credentials.type == "oauth2":
+            oauth_handler = _get_provider_oauth_handler(credentials.provider)
+            if credentials.access_token_expires_at:
+                logger.debug(
+                    f"Credentials #{credentials.id} expire at "
+                    f"{datetime.fromtimestamp(credentials.access_token_expires_at)}; "
+                    f"current time is {datetime.now()}"
+                )
+            else:
+                logger.debug(f"Credentials #{credentials.id} never expire")
+            if oauth_handler.needs_refresh(credentials):
+                logger.debug(
+                    f"Refreshing '{credentials.provider}' credentials #{credentials.id}"
+                )
+                if lock:
+                    # Wait until the credentials are no longer in use anywhere
+                    self._lock(user_id, credentials_id)
+
+                fresh_credentials = oauth_handler.refresh_tokens(credentials)
```

Comment:
Good point. We can just do the check twice (cheap operation but a bit dirty). The alternative would be locking *before* doing the refresh check, but I want to avoid unnecessary locking operations.